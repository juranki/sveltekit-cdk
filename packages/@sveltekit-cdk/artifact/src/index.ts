import { writeFileSync } from 'fs';
import path from 'path';

/**
 * SvelteKitCDKArtifact defines how SvelteKit build
 * artifacts are stored in filesystem.
 */
export class SvelteKitCDKArtifact {
  readonly version = 1;
  readonly metaPath: string;
  readonly staticPath: string;
  readonly lambdaPath: string;

  /**
   * List of glob patterns for routing static content.
   */
  public staticGlobs?: string[];

  constructor(basePath: string) {
    this.staticPath = path.join(basePath, 'static');
    this.lambdaPath = path.join(basePath, 'lambda');
    this.metaPath = path.join(basePath, 'meta.json');
  }

  get subdirectories(): string[] {
    return [
      this.staticPath,
      this.lambdaPath,
    ].copyWithin(3, 0);
  }

  write() {
    writeFileSync(this.metaPath, JSON.stringify({
      version: this.version,
      staticGlobs: this.staticGlobs,
    }, null, 2));
  }
}