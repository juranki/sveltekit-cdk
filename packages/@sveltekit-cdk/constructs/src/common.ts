export const DEFAULT_ARTIFACT_PATH = 'sveltekit';

export interface RendererProps {
  /**
   * Location of sveltekit artifacts
   *
   * @default 'sveltekit'
   */
  readonly artifactPath?: string;
  /**
   * Environment variables for the backend implementation
   */
  readonly environment?: {
    [key: string]: string;
  };
  /**
   * Logging verbosity (default: INFO)
   */
  readonly logLevel?: 'ERROR' | 'WARN' | 'INFO' | 'DEBUG';
}

export type StaticRoutes = Record<string, 'prerendered' | 'static'>
