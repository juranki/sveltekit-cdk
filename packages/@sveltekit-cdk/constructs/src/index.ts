export { RendererProps } from './common';
export { SvelteDistribution, SvelteDistributionProps } from './distribution';
