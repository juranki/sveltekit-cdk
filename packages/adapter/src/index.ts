import type { Adapter } from '@sveltejs/kit'
import * as path from 'path'
import { build } from 'esbuild'

export interface AwsServerlessAdapterParams {
    /**
     * Location of CDK project. Required for automatic deploy.
     */
    cdkProjectPath?: string
    /**
     * Path to store sveltekit artifacts.
     * 
     * @default ${cdkProjectPath}/sveltekit
     */
    artifactPath?: string
    /**
     * Stack to deploy after producing artifact.
     */
    stackName?: string
}

export function AwsServerlessAdapter({
    cdkProjectPath, artifactPath, stackName
}: AwsServerlessAdapterParams): Adapter {
    if (!cdkProjectPath && !artifactPath) {
        throw new Error("at least one of cdkProjectPath or artifactPath is required");
    }
    if (!cdkProjectPath && stackName) {
        throw new Error("when stackName is specified, cdkProjectPath is mandatory");
    }
    return {
        name: 'sveltekit-cdk-adapter',
        async adapt({ utils, config }): Promise<void> {

            const targetPath = artifactPath || path.join(cdkProjectPath!, 'sveltekit')
            const files = path.join(__dirname, 'files');
            const dirs = {
                static: path.join(targetPath, 'static'),
                lambda: path.join(targetPath, 'lambda'),
                server: path.join(targetPath, 'server'),
            }

            utils.rimraf(targetPath)
            utils.rimraf('.svelte-kit/cdk')

            await utils.prerender({
                dest: dirs.static
            });
            utils.copy_client_files(dirs.static)
            utils.copy_static_files(dirs.static)
            utils.copy(path.join(files, 'proxy-handler-v2.ts'), '.svelte-kit/cdk/proxy-handler-v2.ts')

            await build({
                entryPoints: ['.svelte-kit/cdk/proxy-handler-v2.ts'],
                outfile: path.join(dirs.server, 'proxy-handler-v2.js'),
                bundle: true,
                platform: 'node',
                inject: [path.join(files, 'shims.js')]
            })

        },
    }
}