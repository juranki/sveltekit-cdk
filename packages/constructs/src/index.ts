export { SvelteDistribution } from './distribution'
export { SvelteBackendApiV2Lambda } from './backend-apiv2-lambda'
export { SimpleSvelteStack } from './simple-stack'
