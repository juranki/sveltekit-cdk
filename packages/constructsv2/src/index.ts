export { SvelteDistribution, SvelteDistributionProps } from './distribution'
export { RendererProps } from './common'