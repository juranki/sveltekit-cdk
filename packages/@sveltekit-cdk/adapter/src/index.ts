import { writeFileSync, renameSync } from 'fs';
import * as path from 'path';
import type { Adapter, Builder } from '@sveltejs/kit';
import { SvelteKitCDKArtifact } from '@sveltekit-cdk/artifact';
import { build } from 'esbuild';

export interface AdapterParams {
  /**P
   * Location of CDK project.
   *
   * One of `cdkProjectPath` or `artifactPath` is required.
   */
  cdkProjectPath?: string;
  /**
   * Path to store sveltekit artifacts.
   *
   * One of `cdkProjectPath` or `artifactPath` is required.
   *
   * @default ${cdkProjectPath}/sveltekit
   */
  artifactPath?: string;
}

/**
 * Returns adapter that prepares SvelteKit site for deployment with AWS CDK V2
 */
export function adapter({ cdkProjectPath, artifactPath }: AdapterParams): Adapter {
  if (!cdkProjectPath && !artifactPath) {
    throw new Error('at least one of cdkProjectPath or artifactPath is required');
  }
  return {
    name: 'sveltekit-cdk-adapter',
    async adapt(builder): Promise<void> {

      // Prepare code for lambda function
      const lambdaPath = builder.getBuildDirectory('@sveltekit-cdk');
      builder.rimraf(lambdaPath);
      builder.mkdirp(lambdaPath);

      const files = path.join(__dirname, 'files');
      builder.copy(files, lambdaPath, {
        replace: {
          SERVER: '../output/server/index',
          MANIFEST: '../output/server/manifest',
          PRERENDERED: './prerendered',
        },
      });
      writePrerenderedTs(lambdaPath, builder);

      // Prepare artifact for CDK deployment
      const targetPath = artifactPath || path.join(cdkProjectPath!, 'sveltekit');
      builder.rimraf(targetPath);

      const artifact = new SvelteKitCDKArtifact(targetPath);
      artifact.subdirectories.forEach((p) => builder.mkdirp(p));

      builder.writePrerendered(artifact.staticPath);
      const staticFiles = builder.writeClient(artifact.staticPath);
      artifact.staticGlobs = staticGlobs(staticFiles);

      // UGLY WORKAROUND FOR CF/S3 ROUTING FOR FILES WITH + IN PATH
      for (let filename of staticFiles.filter(f => f.includes('+'))) {
        const newFilename = filename.replaceAll('+', ' ');
        renameSync(path.join(artifact.staticPath, filename), path.join(artifact.staticPath, newFilename));
      }

      await build({
        entryPoints: [path.join(lambdaPath, 'at-edge-handler.js')],
        outfile: path.join(artifact.lambdaPath, 'at-edge-handler.js'),
        bundle: true,
        platform: 'node',
        inject: [path.join(lambdaPath, 'shims.js')],
        external: ['./settings'],
      });

      artifact.write();

      builder.log(`CDK artifact was written to ${targetPath}`);
    },
  };
}

/**
 * Returns a list of glob patterns that can be used for
 * creating CloudFront behaviours for serving static content
 *
 * @param staticFiles
 */
function staticGlobs(staticFiles: string[]): string[] {
  const globs: { [glob: string]: boolean } = {};
  staticFiles.forEach(p => {
    const ps = p.split('/');
    const glob = ps.length > 1 ? `${ps[0]}/*` : p;
    globs[glob] = true;
  });
  return Object.keys(globs);
}

function writePrerenderedTs(targetPath: string, builder: Builder) {
  const prerenderedPages: { [route: string]: string } = {};
  builder.prerendered.pages.forEach((v, k) => {
    prerenderedPages[k] = v.file;
  });
  writeFileSync(
    targetPath,
    [
      `export const prerenderedPages = ${JSON.stringify(prerenderedPages)}`,
    ].join('\n'),
  );
}
