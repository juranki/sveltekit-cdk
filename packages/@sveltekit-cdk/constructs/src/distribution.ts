import { randomBytes } from 'crypto';
import { writeFileSync } from 'fs';
import { join } from 'path';
import { SvelteKitCDKArtifact } from '@sveltekit-cdk/artifact';
import { Duration } from 'aws-cdk-lib';
import { Certificate } from 'aws-cdk-lib/aws-certificatemanager';
import * as cdn from 'aws-cdk-lib/aws-cloudfront';
import * as cdnOrigins from 'aws-cdk-lib/aws-cloudfront-origins';
import { CacheControl } from 'aws-cdk-lib/aws-codepipeline-actions';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Construct } from 'constructs';
import { DEFAULT_ARTIFACT_PATH, RendererProps } from './common';

export interface SvelteDistributionProps {
  /**
   * Location of sveltekit artifacts
   *
   * @default 'sveltekit'
   */
  readonly artifactPath?: string;

  /**
   * Props for Lambda@Edge renderer
   */
  readonly rendererProps?: RendererProps;

  /**
   * PriceClass
   *
   * @link https://docs.aws.amazon.com/cdk/api/latest/typescript/api/aws-cloudfront/priceclass.html#aws_cloudfront_PriceClass
   * @default PriceClass.PRICE_CLASS_100
   */
  readonly priceClass?: cdn.PriceClass;
  /**
   * Origin request policy determines which parts of requests
   * CloudFront passes to your backend
   *
   * @default AllViewer managed policy
   * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/controlling-origin-requests.html
   */
  readonly originRequestPolicy?: cdn.IOriginRequestPolicy;

  /**
   * Cache policy determies caching for dynamic content.
   *
   * Note: static content is cached using default setting (CACHING_OPTIMIZED).
   *
   * @default CACHING_DISABLED
   * @link https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/controlling-origin-requests.html
   */
  readonly cachePolicy?: cdn.ICachePolicy;

  /**
   * Bucket props for the svelteStaticBucket s3 bucket.
   *
   * @link https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-s3.BucketProps.html
   */
  readonly bucketProps?: s3.BucketProps;

  /**
   * Certificate to use with the CloudFront Distribution
   *
   * @default undefined
   */
  readonly certificateArn?: string;

  /**
   * Domain names to associate with the CloudFront Distribution
   *
   * @default undefined
   */
  readonly domainNames?: Array<string>;
}

export class SvelteDistribution extends Construct {
  distribution: cdn.Distribution;
  bucket: s3.Bucket;
  function?: cdn.experimental.EdgeFunction;
  constructor(scope: Construct, id: string, props: SvelteDistributionProps) {
    super(scope, id);

    // validate pros and apply defaults
    checkProps(props);
    const artifactPath = props.artifactPath || DEFAULT_ARTIFACT_PATH;
    const artifact = new SvelteKitCDKArtifact(artifactPath);
    artifact.read();
    const envUtils = new EnvUtil(props.rendererProps?.environment || {});


    // origins
    const bucketProps = props.bucketProps || {};
    this.bucket = new s3.Bucket(this, 'svelteStaticBucket', bucketProps);

    const s3origin = new cdnOrigins.S3Origin(this.bucket, {
      originPath: 'static',
    });

    // cache and origin request policies
    const originRequestPolicy = props.originRequestPolicy || cdn.OriginRequestPolicy.ALL_VIEWER;
    const cachePolicy = props.cachePolicy || new cdn.CachePolicy(this, 'svelteDynamicCachePolicy', {
      cookieBehavior: cdn.CacheCookieBehavior.all(),
    });

    writeFileSync(
      join(artifact.lambdaPath, 'settings.js'),
      [
        `export const logLevel = ${JSON.stringify(props.rendererProps?.logLevel || 'INFO')};`,
        `export const headerEnvMap = ${envUtils.mappingJSON()};`,
      ].join('\n'),
    );

    this.function = new cdn.experimental.EdgeFunction(this, 'svelteHandler', {
      code: lambda.Code.fromAsset(artifact.lambdaPath),
      handler: 'at-edge-handler.handler',
      runtime: lambda.Runtime.NODEJS_16_X,
      timeout: Duration.seconds(5),
      memorySize: 512,
      logRetention: 7,
    });

    const edgeLambdas = [{
      eventType: cdn.LambdaEdgeEventType.ORIGIN_REQUEST,
      functionVersion: this.function.currentVersion,
      includeBody: true,
    }];

    // distribution
    this.distribution = new cdn.Distribution(this, 'distro', {
      priceClass: props.priceClass || cdn.PriceClass.PRICE_CLASS_100,
      defaultBehavior: {
        origin: s3origin,
        viewerProtocolPolicy: cdn.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        edgeLambdas,
        allowedMethods: cdn.AllowedMethods.ALLOW_ALL,
        originRequestPolicy,
        cachePolicy,
      },
      domainNames: props.domainNames ? props.domainNames : undefined,
      certificate: props.certificateArn ? Certificate.fromCertificateArn(this, 'domainCert', props.certificateArn) : undefined,
    });

    // routes for static content
    artifact.staticGlobs.forEach((glob) => {
      this.distribution.addBehavior(glob, s3origin, {
        viewerProtocolPolicy: cdn.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      });
    });

    new BucketDeployment(this, 'svelteStaticDeployment', {
      destinationBucket: this.bucket,
      destinationKeyPrefix: 'static',
      sources: [Source.asset(artifact.staticPath)],
      distribution: this.distribution,
      cacheControl: [
        CacheControl.maxAge(Duration.days(365)),
      ],
      distributionPaths: getDistributionPaths(artifact),
    });
  }
}

function checkProps(props: SvelteDistributionProps) {
  if (props.certificateArn && !props.domainNames) {
    throw new Error('domainNames must be provided when setting a certificateArn');
  }
}

/**
 * Prepare structures for passing env vars into Lambda@Edge
 */
class EnvUtil {
  values: Record<string, string>;
  headers: Record<string, string>;

  constructor(env: Record<string, string>) {
    this.values = env;
    this.headers = Object.fromEntries(
      Object.keys(env).map(k => [k, `x-env-${randomBytes(9).toString('hex')}`]),
    );
  }

  mappingJSON = () => {
    return JSON.stringify(Object.fromEntries(
      Object.entries(this.headers).map(([k, v]) => ([v, k]))));
  };

  customHeaders = () => {
    return Object.fromEntries(Object.entries(this.values).map(([k, v]) => (
      [this.headers[k], v]
    )));
  };
}

function getDistributionPaths(artifact: SvelteKitCDKArtifact): string[] {
  return artifact.prerenderedRoutes.concat(artifact.staticGlobs).map(p => {
    if (p.startsWith('/')) {
      return p;
    } else {
      return `/${p}`;
    };
  });
}

