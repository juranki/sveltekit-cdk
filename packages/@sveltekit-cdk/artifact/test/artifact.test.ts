import { mkdirSync, readFileSync } from 'fs';
import { SvelteKitCDKArtifact } from '../src';

test('new artifact', () => {
  const artifact = new SvelteKitCDKArtifact('dummy');
  const subs = artifact.subdirectories;
  expect(subs.length).toBe(2);
  expect(subs).toContain('dummy/static');
  expect(subs).toContain('dummy/lambda');
});

test('write', () => {
  const artifact = new SvelteKitCDKArtifact('test-data');
  mkdirSync('test-data', {
    recursive: true,
  });
  artifact.staticGlobs = ['/', 'favicon.ico', '_app/*'];
  artifact.write();
  const json = readFileSync('test-data/meta.json', { encoding: 'utf-8' });
  expect(json).toMatchSnapshot();
});