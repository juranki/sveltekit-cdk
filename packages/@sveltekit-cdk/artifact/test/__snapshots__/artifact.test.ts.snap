// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write 1`] = `
"{
  "version": 1,
  "staticGlobs": [
    "/",
    "favicon.ico",
    "_app/*"
  ]
}"
`;
